File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter2/Section12/CapturePriceFiltered.cs
using System;

using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter2.Section12
{
  public class CapturePriceFiltered : IPump
  {
    public Outputs Create(Inputs inputs)
    {
      return new Outputs()
        .SetPriceLcd1(perFuel(inputs.ENozzle1, inputs.Price1))
        .SetPriceLcd2(perFuel(inputs.ENozzle2, inputs.Price2))
        .SetPriceLcd3(perFuel(inputs.ENozzle3, inputs.Price3));
    }

    enum StartFill
    {
      START_FILL
    };

    static Behavior<String> perFuel(
      Event<UpDown> eNozzle,
      Behavior<Double> price)
    {
      Event<StartFill> eStartFill = eNozzle.Filter(u => u == UpDown.UP) .Map(u => StartFill.START_FILL);
      Behavior<Double> capPrice = eStartFill.Snapshot(price).Hold(0.0);
      Behavior<UpDown> nozzle = eNozzle.Hold(UpDown.DOWN);
      return Behavior<UpDown>.Lift(
        (u, price_) => u.Equals(UpDown.UP) ? Formatters.FormatPrice(price_) : "",
        nozzle,
        capPrice);
    }
  }
}

File: book/petrol-pump/java/chapter2/section12/CapturePriceFiltered.java
package chapter2.section12;

import pump.*;
import sodium.*;

public class CapturePriceFiltered implements Pump
{
    public Outputs create(Inputs inputs) {
        return new Outputs()
            .setPriceLCD1(perFuel(inputs.sNozzle1, inputs.price1))
            .setPriceLCD2(perFuel(inputs.sNozzle2, inputs.price2))
            .setPriceLCD3(perFuel(inputs.sNozzle3, inputs.price3));
    }

    enum StartFill { START_FILL };

    private static Cell<String> perFuel(
                Stream<UpDown> sNozzle, Cell<Double> price) {
        Stream<StartFill> sStartFill =
                               sNozzle.filter(u -> u == UpDown.UP)
                                      .map(u -> StartFill.START_FILL);
        Cell<Double> capPrice = sStartFill.snapshot(price).hold(0.0);
        Cell<UpDown> nozzle = sNozzle.hold(UpDown.DOWN);
        return Cell.lift(
            (u, price_) ->
                u.equals(UpDown.UP) ? Formatters.formatPrice(price_)
                                    : "",
            nozzle, capPrice);
    }
}

